src/java/picard/analysis/directed/TargetMetricsCollector.java

```diff
@@ -425,32 +590,51 @@ private void calculateTargetCoverageMetrics() {
+            final int targetBasesDepth[] = {0, 1, 2, 10, 20, 30, 40, 50, 100}; // NB: this should be in ascending order
+            final int targetBases[] = new int[targetBasesDepth.length]; // counts for how many target bases are at at least X coverage, where X corresponds to the value at the same offset in targetBasesDepth
 
+            // consider all bases in calculating the mean, median etc.
+            for (final Coverage c : this.coverageByTarget.values()) {
+                final boolean hasCoverage = c.hasCoverage();
                 final int[] targetDepths = c.getDepths();
-                basesConsidered += targetDepths.length;
+
+                if (!hasCoverage) zeroCoverageTargets++;
 
                 for (final int depth : targetDepths) {
-                    depths[depthIndex++] = depth;
-                    totalCoverage += depth;
+                    targetCoverageHistogram.increment(depth);
+                    if (0 == depth) numTargetBasesAtZeroCoverage++;
+                    else totalCoverage += depth;
+                    if (hasCoverage) depths[depthIndex++] = depth;
+
+                    // Add to the "how many bases at at-least X" calculations.
+                    for (int i = 0; i < targetBasesDepth.length; i++) {
+                        if (depth >= targetBasesDepth[i]) targetBases[i]++;
+                        else break; // NB: assumes that targetBasesDepth is sorted in ascending order
+                    }
                 }
             }
 
-            this.metrics.MEAN_TARGET_COVERAGE = totalCoverage / basesConsidered;
+            // Sort the array (ASCENDING)
+            Arrays.sort(depths);
+
+            this.metrics.MEAN_TARGET_COVERAGE = totalCoverage / this.metrics.TARGET_TERRITORY;
+            this.metrics.MEDIAN_TARGET_COVERAGE =
+                    (0 == (depths.length % 2)) ? ((depths[(depths.length/2) - 1] + depths[(depths.length/2)]) / 2.0) : depths[(depths.length-1)/2];
 
-            // Sort the array (ASCENDING) and then find the base the coverage value that lies at the 80%
+
+            // get the number of bases at zero coverage, and subtract that from the total coverage, then divide that by the total coverage
+            // to get the percent of target bases at 1x coverage or greater.
+            this.metrics.PCT_TARGET_BASES_1X = (this.metrics.TARGET_TERRITORY - numTargetBasesAtZeroCoverage) / this.metrics.TARGET_TERRITORY;
+
+            // find the base the coverage value that lies at the 80%
             // line, which is actually at 20% into the array now
-            Arrays.sort(depths);
-            // Note.  basesConsidered can be between 0 and depths.length inclusive.  indexOf80thPercentile will be -1 in the latter case
-            final int indexOf80thPercentile = Math.max((depths.length - 1 - basesConsidered) + (int) (basesConsidered * 0.2), 0);
+            final int indexOf80thPercentile = Math.max((int) (this.metrics.TARGET_TERRITORY * 0.2 ) - 1, 0);
```

Comment:
Given that you didn't include the depth from all the !hasCoverage targets in this calculation (nor do you want to) I think you should use depthIndex rather than TARGET_TERRITORY for this. Am I missing something?